### PR TITLE
DNN: optimize dnn vulkan backend

### PR DESCRIPTION
**Pull request by zihaomu**
## Optimize DNN Vulkan backend

merge with: https://github.com/opencv/ci-gha-workflow/pull/95.

My purposes for this PR:
1. upgrade the Vulkan header file from version `1.0` to `1.2` to support the `fp16` and `int8` data format.
2. Carefully optimized the convolution layer and gemm layer. speed up from 170 ms to 36 ms of ResNet50 with Vulkan Backend.
3. Remove support for some layers like: pooling, permute, LRN, relu. The support of these layers will slow down the DNN inference speed because their kernels are not well-optimized. I think you should leave this task for the next step. GSoC students could take on some work.
4. Support the ios and Mac M1 chip platforms.

## Vulkan CI result can be found at [this PR](https://github.com/zihaomu/opencv/pull/1)
We only optimize the integrated GPU, and the discrete GPU like Nvidia GPU will run relatively slowly.
There are two CIs:
1. Mac M1, running the full test would take about 2 mins.
2. Win10, Nvidia GPU, running the full test would take about 5 mins.

## TODO List: 
- add the vulkan CI in github action, then we can test the PR.

## Performance Test
NOTE: Currently PR is only optimized for integrated graphics, it will run very slowly on discrete graphics like Nvidia GPU. 

Test on Apple M1 chip.

| Model Name | Resnet50 | MobileNetV2 | YoloV3  | YoloV4 |
| --------------------- | --------- | ---------- | ------ | ------ |
| CPU Backend (4 thread) | 26 ms   | 6 ms     | 130.05 ms |215.76 ms|
| CPU without Winograd (4 thread) | 35 ms | 6.5 ms | 218.9 ms |271.7 ms|
| Vulkan GPU | 37.8 ms | 13.8 ms | 182.3 s |270.04 ms|

Patch performance:
Since the old vulkan kernel is almost without optimize, it works very slowly.

| Test of ResNet50 on M1 chip | Before patch | With patch |
| --------------------- | --------- | ---------- |
| Vulkan GPU | 190 ms | **37.8 ms (4X faster)**|

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@vpisarev Friendly reminder.

**Comment by zihaomu** [comment]
The CI is greed now. https://github.com/zihaomu/opencv/pull/1

File: modules/dnn/src/int8layers/pooling_layer.cpp [hunk 1]
```diff
@@ -89,10 +89,10 @@ class PoolingLayerInt8Impl CV_FINAL : public PoolingLayerInt8
...
         {
             // Pool1D
-            kernel_size.assign(1, kernel_size[0]);
-            strides.assign(1, strides[0]);
-            pads_begin.assign(1, pads_begin[0]);
-            pads_end.assign(1, pads_end[0]);
+            kernel_size.resize(1, kernel_size[0]);
+            strides.resize(1, strides[0]);
+            pads_begin.resize(1, pads_begin[0]);
+            pads_end.resize(1, pads_end[0]);
```
Comment (zihaomu) [inline modules/dnn/src/int8layers/pooling_layer.cpp:95]: This modification fixes the error reported by Visual Studio 2020.

**Comment by opencv-alalek** [comment]
Please rebase to resolve conflicts:

```
Conflicting files
modules/dnn/src/dnn_common.hpp
modules/dnn/test/test_backends.cpp
```

**Pull request merged**